### PR TITLE
fix: stop destroying a registry.json that could only not be read, and stop orphaning failed downloads

### LINKED ISSUE #323: load_registry destroys registry.json when it merely could not be READ
Found by the pre-push gate on the #314 branch, by two independent reviewers. Pre-existing, and deliberately not fixed there: the fix changes behaviour on a recovery path and wants its own test matrix.

`load_registry` treats every failure to LOAD `registry.json` as grounds to overwrite it with the bundled copy.

```rust
let user_registry = match load_from_file(&registry_path) {
    Ok(registry) => registry,
    Err(e) => {
        tracing::warn!("Failed to load user registry at {}: {}. Using bundled registry.", ...);
        write_registry_file(&registry_path, &bundled_registry)?;   // <- destroys the file
        return Ok(bundled_registry);
    }
};
```

`load_from_file` returns two semantically different errors and the branch cannot tell them apart:

- `Error::RegistryParse`, the file is definitively corrupt. Overwriting is right.
- `Error::RegistryRead`, an I/O failure. EACCES, EIO, EMFILE, a flaky network mount, fd exhaustion during a parallel batch run. The file may be perfectly intact and simply unreadable at that instant.

Only the first justifies replacing it. In the second the tool destroys a file it could not even read, which is the "could not determine, so assume the destructive answer" shape that has turned up several times in this backlog now.

The cost is real rather than theoretical: any user-local edits to `registry.json` are gone with no message, and birda-gui reads the file straight off disk, so a shared registry someone maintained by hand is replaced silently.

## Fix

Three outcomes rather than two:

- `RegistryParse` -> overwrite with the bundled registry, as today.
- `RegistryRead` -> use the bundled registry **in memory**, leave the file alone, and say so in the warning.

About three lines, plus a test for each arm. The test matrix is why this is not folded into #314: the read arm needs a fixture that makes reading fail without making the file corrupt (a mode-0000 file with a root guard, or a directory in its place), and the parse arm must keep the existing overwrite behaviour pinned.

## Related

- #311 (`registry.json` written non-atomically) is fixed by the same PR that found this, so the recovery write is now atomic. That caps the damage but does not remove it: an intact-but-unreadable registry is still replaced.
- Same class as #309 item 3 and the config-repair work in #305, one artifact over.

### PR DESCRIPTION
**Pull request by tphakala**
Closes #323.

Relates to #342: this fixes item 1 of that issue (a failed `finalize_download` orphaning its `.part` file) and nothing else. Items 2, 3 and 4 stay open, so #342 is deliberately not closed here. Both changes are the same defect class in the same module, `src/registry/`: a failure path that took the destructive or wasteful branch because it could not tell two situations apart.

Two failure paths that took the destructive or wasteful branch because they could not tell two situations apart.

## #323: a registry that could not be READ was replaced

`load_registry` treated every failure out of `load_from_file` as grounds to overwrite `registry.json` with the bundled copy. That function returns two semantically different errors and the arm could not distinguish them: `RegistryParse` means the file is not a registry, where replacing it is the repair, but `RegistryRead` is an I/O failure (EACCES, EIO, EMFILE, a flaky network mount, fd exhaustion during a parallel batch) that says nothing about the contents. The file was very possibly intact and was destroyed anyway, taking any local edits with it, with a warning as the only trace. birda-gui reads this file straight off disk, so a hand-maintained registry disappearing is user visible.

The parse arm keeps today behaviour. Everything else now uses the bundled registry in memory and leaves the file alone, so an error variant added later defaults to the non-destructive branch rather than the destructive one.

## #342 item 1: a failed publish orphaned its part file

`download_verified` has three failure paths after the temp file exists. The stream-failure and checksum-failure paths both remove it; the rename path propagated with no cleanup. `part_path` qualifies the name with the process id, so the next attempt picks a different name and never reclaims the old one, and `find_obsolete_files` matches a fixed list of names that does not include it. Model files are tens to hundreds of MB, so a user whose disk fills, or who hits the documented EBUSY on a bind-mounted destination, silently lost that much space per attempt with no way to find it short of `find`. Cleanup now runs through `roll_back`, which this file already uses for the same job and which reports the path when the unlink itself fails.

## A regression the review caught, and the fix for it

The first version of the #323 fix split on the error variant, but `read_to_string` validates UTF-8 and reports a failure as `io::ErrorKind::InvalidData`, so it arrived as `RegistryRead`. That put a definitively corrupt registry on the preserve side: never repaired, never upgraded, warning on every run, no route back but deleting the file by hand. The triggers are ordinary rather than exotic. Notepad "Unicode" and PowerShell 5.1 `>` both write UTF-16; a cp1252 round trip mangles the a-umlaut the bundled registry already carries; a pre-atomic-write birda killed mid-write could truncate one mid-codepoint.

Reading bytes and parsing bytes puts the split on the contents-versus-transport axis with no special case, so `RegistryRead` now means only a failure to obtain the bytes. One measured limit is stated in the code: `from_slice` does not validate UTF-8 inside a field the deserializer ignores, which is the harmless direction (such a file is accepted, not destroyed).

## Warnings

All three warnings in the file now render the error and its cause instead of a hand-built string, so the path appears once and the errno appears at all. `thiserror` puts only the `#[error(..)]` text into `Display`, so previously "could not be read" was the whole message and EACCES, EIO and EMFILE, which want three different responses, were indistinguishable. Verified against the built binary:

```
failed to read registry file X: Permission denied (os error 13). Continuing with the bundled registry; the file on disk is left as it is and will be read again on the next run.
failed to parse registry file X: key must be a string at line 1 column 3. Replacing it with the bundled registry.
failed to write registry file X: Permission denied (os error 13). Continuing with the bundled registry; this will be retried on the next run.
```

## Behaviour change worth a second opinion

Against `main` there is no on-disk state where this branch is worse; every state it replaces, `main` replaced too. Against the intermediate commit in this branch, a structurally intact hand-maintained registry in UTF-16 or cp1252 is now replaced rather than preserved. That is deliberate (those bytes are not JSON for any consumer, birda-gui included) and is documented in the README, but it is the one square where reasonable people could want the other answer.

Separately, rendering the parse error cause means serde can quote a value from the file into a warning, for example `invalid type: string "..." , expected u32`. `registry.json` is the model catalogue rather than a secret store, but it is hand-edited and private mirrors are supported, so it is worth knowing that stderr can now carry a fragment of the file.

## Tests

The two behavioural fixes were written test-first and each new test was confirmed to fail against the pre-fix code before the fix landed. The UTF-16 class is pinned by a test that is deliberately not `cfg(unix)`, because the ways to produce it are mostly Windows ones and CI runs tests on Linux only (#315). The Windows rename-onto-a-directory behaviour the installer test rests on was verified on a real Windows 11 VM (`MoveFileExW` with `MOVEFILE_REPLACE_EXISTING` returns `ERROR_ACCESS_DENIED`) rather than inferred, so the test is portable even though nothing runs it there.

Also pinned, after mutation testing showed they survived: the equal-version boundary, a keep path that rewrites the file anyway, the first-run bootstrap write, an error naming the part file instead of the destination, and `cause_of` itself.

`cargo fmt --check`, `cargo clippy --no-default-features --all-targets -- -D warnings` and `cargo test --no-default-features` are clean; 658 tests pass.

## Filed rather than fixed here

Pre-existing findings from the same review, all out of scope: #347, #348, #349, #350 and the batched #351.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved registry recovery when cached files are unreadable, corrupted, invalidly encoded, or outdated.
  * Preserved unreadable registry files while using the bundled registry for the current run.
  * Added clearer warnings that include the underlying cause of registry loading or persistence failures.
  * Cleaned up temporary download files when installation publication fails.
  * Ensured failed downloads report the correct installation destination and can be retried safely.

### CODE REVIEW

**Comment by tphakala** [comment]
On the out-of-scope check: the download cleanup is item 1 of #342, and I have made that link explicit in the description rather than splitting the PR. Both changes are the same defect in the same module, `src/registry/`, where a failure path takes the destructive or wasteful branch because it cannot tell two situations apart, and both are a handful of lines. I did not write `Closes #342` because items 2, 3 and 4 of that issue are untouched and it should stay open.

File: src/registry/loader.rs [hunk 1]
```diff
@@ -6,31 +6,78 @@ use std::path::PathBuf;
...
+/// writes the developer's real config directory.
+///
+/// Infallible on purpose: every failure below is already a degradation to the
+/// bundled registry, which is in hand before this is called, so there is nothing
+/// left that could justify aborting the command that asked for it.
+fn load_registry_from(registry_path: &std::path::Path, bundled_registry: Registry) -> Registry {
     if !registry_path.exists() {
-        return Ok(bootstrap_registry(&registry_path, &bundled_registry));
+        return bootstrap_registry(registry_path, &bundled_registry);
     }
```
[the remark this answers is not in the document]
Comment (tphakala) [reply src/registry/loader.rs:65]: Correct, and fixed in 99e5111. I had found this in my own pre-push review and filed it as #347 rather than fixing it, on the grounds that it was pre-existing; you are right that leaving it makes the PR internally inconsistent, since it is the same collapse three lines above the one being fixed.

`try_exists()` now supplies the third outcome. Only that outcome changes: `Ok(true)` and `Ok(false)` behave exactly as before, including a dangling symlink, which reports `Ok(false)` and is still bootstrapped over.

One detail worth recording, because it is why there is no test. The reachable permission case was already self-protecting: an EACCES that breaks the `stat` breaks the following write too, so the old code survived by the write failing rather than by the check being right. Reaching the destructive outcome needs a failure that breaks `stat` on the inode but not create-and-rename in its directory, such as ESTALE after an NFS server restart, which no portable fixture produces. A test against the reachable case would pass before and after. Verified by hand instead: a mode-0000 config directory now reports `Could not determine whether the model registry at X exists: Permission denied (os error 13). Continuing with the bundled registry; nothing on disk is touched.`
